ouroboros-consensus/test-consensus/Test/Dynamic/RealPBFT.hs

```diff
@@ -0,0 +1,141 @@
+    ]
+  where
+    sp = defaultSecurityParam
+
+prop_simple_real_pbft_convergence :: SecurityParam
+                                  -> NumCoreNodes
+                                  -> NumSlots
+                                  -> Seed
+                                  -> Property
+prop_simple_real_pbft_convergence sp numCoreNodes =
+    give (Genesis.gdProtocolMagicId $ Genesis.configGenesisData genesisConfig) $
+    give (Genesis.configEpochSlots genesisConfig) $
+    prop_simple_protocol_convergence
+      (\cid -> protocolInfo numCoreNodes cid
+                            (mkProtocolRealPBFT numCoreNodes cid genesisConfig))
+      isValid
+      numCoreNodes
+  where
+    genesisConfig = mkGenesisConfig numCoreNodes
+
+    isValid :: ByronGiven
+            => [NodeId]
+            -> TestOutput (ByronBlockOrEBB ByronConfig)
+            -> Property
+    isValid nodeIds TestOutput{testOutputNodes = final} =
+          counterexample (show final')
+     $    tabulate "shortestLength" [show (rangeK sp (shortestLength final'))]
+     $    Map.keys final === nodeIds
+     .&&. allEqual (takeChainPrefix <$> Map.elems final')
+     .&&. not (all Chain.null $ Map.elems final')
+      where
+        -- Without the 'NodeConfig's
+        final' = snd <$> final
+        takeChainPrefix :: Chain (ByronBlockOrEBB ByronConfig)
+                        -> Chain (ByronBlockOrEBB ByronConfig)
+        takeChainPrefix = id -- in PBFT, chains should indeed all be equal.
+
+
+mkProtocolRealPBFT :: NumCoreNodes
+                   -> CoreNodeId
+                   -> Genesis.Config
+                   -> Protocol (ByronBlockOrEBB ByronConfig)
+mkProtocolRealPBFT (NumCoreNodes n) (CoreNodeId i) genesisConfig =
+    ProtocolRealPBFT
+      genesisConfig
+      (Just signatureThreshold)
+      (Update.ProtocolVersion 1 0 0)
+      (Update.SoftwareVersion (Update.ApplicationName "Cardano Test") 2)
+      (Just leaderCredentials)
+  where
+    leaderCredentials :: PBftLeaderCredentials
+    leaderCredentials = either (error . show) id $
+        mkPBftLeaderCredentials
+          genesisConfig
+          dlgKey
+          dlgCert
+
+    signatureThreshold = PBftSignatureThreshold $
+      1.0 / (fromIntegral n + 1.0)
+
+    dlgKey :: Crypto.SigningKey
+    dlgKey = (!! i)
+           . Genesis.gsRichSecrets
+           . fromJust
+           $ Genesis.configGeneratedSecrets genesisConfig
+
+    dlgCert :: Delegation.Certificate
+    dlgCert = fromJust $
+      find (\crt -> Delegation.delegateVK crt == Crypto.toVerification dlgKey)
+           (Map.elems dlgMap)
+
+    dlgMap :: Map Common.KeyHash Delegation.Certificate
+    dlgMap = Genesis.unGenesisDelegation
+           $ Genesis.gdHeavyDelegation
+           $ Genesis.configGenesisData genesisConfig
+
+
+-- Instead of using 'Dummy.dummyConfig', which hard codes the number of rich
+-- men (= CoreNodes for us) to 4, we generate a dummy config with the given
+-- number of rich men.
```

Comment:
Did this go wrong in the case where the rich men were too few?

Reply:
Indeed, with `Dummy.dummyConfig`:
```
*** Failed! Exception: 'Prelude.!!: index too large' (after 9 tests and 6 shrinks):
NumCoreNodes 5
...
```